pkg/storage/tsdb/bucketindex/storage.go

```diff
@@ -20,14 +21,76 @@ import (
 )
 
 var (
-	ErrIndexNotFound  = errors.New("bucket index not found")
-	ErrIndexCorrupted = errors.New("bucket index corrupted")
+	ErrIndexNotFound    = errors.New("bucket index not found")
+	ErrIndexCorrupted   = errors.New("bucket index corrupted")
+	ErrIndexNotModified = errors.New("bucket index not modified")
 )
 
+// ReadNewIndex reads bucket index only if it has been modified since lastModified, otherwise it returns ErrIndexNotModified
+func ReadNewIndex(
+	ctx context.Context,
+	bkt objstore.Bucket,
+	userID string,
+	cfgProvider bucket.TenantConfigProvider,
+	logger log.Logger,
+	lastModified time.Time,
+) (*IndexWithLastModified, error) {
+	userBkt := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+	attrs, err := userBkt.WithExpectedErrs(userBkt.IsObjNotFoundErr).Attributes(ctx, IndexCompressedFilename)
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return nil, ErrIndexNotFound
+		}
+		return nil, errors.Wrap(err, "attribute bucket index")
+	}
+	if !attrs.LastModified.After(lastModified) {
+		return nil, ErrIndexNotModified
+	}
+	idx, err := readIndex(ctx, userBkt, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &IndexWithLastModified{
+		Index:        idx,
+		LastModified: attrs.LastModified,
```

Comment:
Just wanted to add a note about something I noticed while giving this a quick glance: this is mixing the last modified time from the HEAD and the result of the GET, which is potentially confusing and could cause extra reads when they're not consistent. Looks like this is a limitation of the `BucketReader` interface.